### ISSUE TITLE
Input to model

### ISSUE THREAD

**Issue by rmmal** [Dobiasd/frugally-deep#6]
If I have  RGB Image
and i want it to  pass it to the model , what should i do ?

what i've made is flatten the input image into vector of float , i appened the r , g , b values after each others to get just 1 vector called "input_vector" 

and then this is the next step.

```
         typedef fplus::shared_ref<std::vector<float>> shared_float_vec;
         shared_float_vec x(fplus::make_shared_ref<vector<float>>(std::move(input_vector)));
         const auto result = decision_model.predict({fdeep::tensor3(fdeep::shape3(3,60,60),x)});
```

then the output is incorrect , what should i do then ? or what i've done wrong ?

**Comment by Dobiasd** [comment]
You mean the output is different from the prediction of the same model in Keras? In that case the input you give to the model in C++ is probably different from what you feed into it in Python.

Did you make sure RGB really is the correct byte order and not BGR?

Is the scaling of the values correct? Often RGB images are loaded into bytes (`0 to 255`) but float value images (often used in training of Keras models) are in the range of `0.0 to 1.0`.

In case your image is loaded into bytes, you could use [`fdeep::tensor3_from_bytes`](https://github.com/Dobiasd/frugally-deep/blob/d0b8cd2f07b0efd8792cd1b234e031d573e23844/include/fdeep/tensor3.hpp#L423) instead of constructing the `shared_float_vec` on your own.

---

If all this does help: Can you perhaps create a minimal example I can use to reproduce the problem?

**Comment by Dobiasd** [comment]
Hi @rmmal,

did you manage to fix the problem, or can I help in any other way?

**Comment by Dobiasd** [comment]
Closed due to inactivity. Just let me know in case the problem still exists, so we can re-open.

**Comment by Ihsan149** [comment]
I modified a bit your library for using my costume loss(dice) and convert a segmentation model. I got a json file.
Can you please tell how we can use give an image for segmentation?

**Comment by johndeppe** [comment]
I'm also getting started with frugally-deep. A toy image-input example would be greatly appreciated.

**Comment by Dobiasd** [comment]
@Ihsan149 As mentioned in the [other issue](https://github.com/Dobiasd/frugally-deep/issues/1#issuecomment-378666475): you can use [`fdeep::tensor3_from_bytes`](https://github.com/Dobiasd/frugally-deep/blob/cc2b4904e80b8a3cc0c08afd8a5753fbac907905/include/fdeep/tensor3.hpp#L416) to convert byte pixel values into an `fdeep::tensor3` to use in `model::predict`.

The library should not need to know (or care) if the output values of the model you use represent a segmentation or something else.

**Comment by Dobiasd** [comment]
@johndeppe If you have the pixels of your image loaded into memory, it's easy to use them for [`fdeep::tensor3_from_bytes`](https://github.com/Dobiasd/frugally-deep/blob/cc2b4904e80b8a3cc0c08afd8a5753fbac907905/include/fdeep/tensor3.hpp#L416).
However way the image is loaded depends on the image library you use (OpenCV, CImg, etc.). If you name your image library I can quickly write up an example code using it for you.

**Comment by johndeppe** [comment]
Thank you!

My problem is trying to pick an image library. I'm cross-compiling for/on an embedded system (a Cyclone V for a school project) and so frugally-deep's header-only nature is ideal. I haven't had much luck finding a similar header-only library for images yet.

**Comment by johndeppe** [comment]
So I guess what I am saying is that I am looking for a suggestion on what image library fits with frugally-deep's philosophy and maybe a toy example for it.

**Comment by Dobiasd** [comment]
Have you checked out CImg yet?
`http://cimg.eu/` writes:
> CImg is lightweight. It is made of a single header file CImg.h that must be included in your C++ source.

**Comment by johndeppe** [comment]
That looks pretty good! I hadn't found that.

Navigating the channels_first /  channels_last hazard is another bit I'm worried about. I will give CImg a go. I would still really appreciate that quick example you mentioned.

**Comment by Dobiasd** [comment]
Ok, just created one example using [CImg](http://cimg.eu/) and one using [OpenCV](https://opencv.org/):
- [cimg_example.cpp](https://gist.github.com/Dobiasd/21651861b73042762126e8eea52d9974)
- [opencv_example.cpp](https://gist.github.com/Dobiasd/3140cfd9f539b6adb346e0b4a0ce157b)

**Comment by johndeppe** [comment]
Thank you! This was exactly what I needed.

Now I just have to figure out why my classifier is always outputting 0.

**Comment by Ihsan149** [comment]
@Dobiasd segmentation model mean, that I converted UNet CNN model , which is used for segmentation ( pixel wise prediction).

**Comment by Dobiasd** [comment]
@johndeppe Assuming your model classified correctly when used in Python, make sure the input you feed to it in C++ is exactly the same. If you do any image-preprocessing steps in Python (e.g. skaling the pixel values into a certain value range etc.) before calling predict, you have to do the same in C++.

You can verify this by simply printing out some pixel values you feed to `predict` in Python and C++ and check for equality.

**Comment by Ihsan149** [comment]
The folowing code is in python.
frcnn_Bbx = cv2.cvtColor(image, cv2.COLOR_BGR2GRAY)
frcnn_Bbx = np.ndarray((1, 160, 160))
frcnn_Bbx =np.expand_dims(frcnn_Bbx,axis=3)
preds=unet_model.predict(frcnn_Bbx )

and this is what I am doing after conversion
    const Mat image = imread("11.png");
    imshow("Input Image",image);
    cvtColor(image, image,cv::COLOR_BayerRG2GRAY);
    assert(image.isContinuous());
    const auto model = fdeep::load_model("fdeep_model.json");
    // Use the correct scaling, i.e. low and high.
    const auto input = fdeep::tensor3_from_bytes(image.ptr(),image.rows, image.cols, image.channels(), 0.0f, 1.0f);
    const auto result = model.predict({input});
    //cout <<"Result:"<< result << endl;
    imshow("result",result);
    imwrite("result.png",result);
and i get this error.
/usr/local/include/opencv2/core/traits.hpp:366: error: ‘type’ is not a member of ‘cv::DataType<fdeep::internal::tensor3>’
 { enum { value = DataType<T>::type }; };
        ^

**Comment by Ihsan149** [comment]
@Dobiasd  
3160 x 160Parsing json ... done. elapsed time: 0.825130 s
Running test 1 of 1 ... terminate called after throwing an instance of 'std::runtime_error'
  what():  not yet implemented
Aborted (core dumped)
I am getting this error.
Can you please have a look at it.
Thank you!

**Comment by Dobiasd** [comment]
`model.predict` returns an `fdeep::tensor3`.
`cv::imshow` and `cv::imwrite` however take `cv::Mat`.

If cou create a `cv::Mat` with the correct height, width, depth and data type (`8U`) you can use `fdeep::tensor3_into_bytes` to copy/convert the values from your `fdeep::tensor3` into it.

**Comment by Dobiasd** [comment]
Regarding your `runtime_error`: 

If you upload your model (`h5` and `json`) I can download it and try to reproduce the error and search for the cause.

**Comment by Ihsan149** [comment]
[https://drive.google.com/drive/folders/1RlA3X4j5tDHmCWDzh-jhpFEl_xFf-k4a?usp=sharing](url)
I uploaded the h5,json and a test image to this gdrive link.

**Comment by Dobiasd** [comment]
First the smaller thing: Using `COLOR_BayerRG2GRAY` does not make sense for the image `11.png` you uploaded. This already is a one-channel image, so after `cv::imread` you don't need to convert it to anything, since your model takes 160x160x1 tensors as input.

OK, but now to the main problem: [`Conv2DTranspose` is not yet supported/implemented](https://github.com/Dobiasd/frugally-deep/issues/26). I just [made a small change](https://github.com/Dobiasd/frugally-deep/commit/272791800d237f5636d85a68bf32183722170048) so that the library now provides a better error message in a case like yours.

However, of course I understand that this does not solve your problem, since you need `Conv2DTranspose` to run your model. So I'm sorry, but currently (i.e. as long as `Conv2DTranspose` is not implemented) you can not use your model with frugally-deep. Perhaps the C++ interface of TensorFlow is a valid alternative for you.

**Comment by Ihsan149** [comment]
@Dobiasd  Thank you for the clarification.
I can use UpSampling2D instead of Conv2DTranspose. Is frugally-deep library support UpSampling2D?

**Comment by Dobiasd** [comment]
Yes, it does. :)

**Comment by Ihsan149** [comment]
@Dobiasd  can you please upload the test program to frugally-deep library that runs the model and gives back an image  instead of fdeep::tensor3  to visualize it using opencv.

**Comment by Dobiasd** [comment]
I've [added](https://github.com/Dobiasd/frugally-deep/commit/7d592b23e0124413f702db1f6a2e73d00320d721) an [example](https://gist.github.com/Dobiasd/7ef20a0ad47d3f8dc1654a0ca5d1c77c) demonstrating not only how to convert a `cv::Mat` to an `fdeep::tensor3`, but also how to convert an `fdeep::tensor3` to a `cv::Mat`. The second part is what you need to be able to visualize the result of your model.

**Comment by Dobiasd** [comment]
You are not using the result of `model.predict({tensor});` but just displaying your input tensor instead. ;)

**Comment by Ihsan149** [comment]
When  i try to use the result of `model.predict({tensor});`
`const cv::Mat image1 = cv::imread("11.png",CV_LOAD_IMAGE_GRAYSCALE);

    const auto model = fdeep::load_model("fdeep_model.json");

    // convert cv::Mat to fdeep::tensor3 (image1 to tensor)
    const fdeep::tensor3 tensor = fdeep::tensor3_from_bytes(image1.ptr(),image1.rows, image1.cols, image1.channels());

    // choose the correct pixel type for cv::Mat (gray or RGB/BGR)
    assert(tensor.shape().depth_ == 1 || tensor.shape().depth_ == 3);
    const int mat_type = tensor.shape().depth_ == 1 ? CV_8UC1 : CV_8UC3;

    //Prediction model,give you tensor as input
    const auto output_tensor = model.predict({tensor});

    const cv::Mat image2(output_tensor.data()->shape().width_,output_tensor.data()->shape().height_,mat_type);

    //const cv::Mat image2(cv::Size(tensor.shape().width_, tensor.shape().height_), mat_type);
    //fdeep::tensor3_into_bytes(input,image2.data)
    fdeep::tensor3_into_bytes(output_tensor,image2.data, image2.rows * image2.cols * image2.channels());
    cv::imshow("image2", image2);
    cv::waitKey();`

 i get this error
error: invalid initialization of reference of type ‘const tensor3& {aka const fdeep::internal::tensor3&}’ from expression of type ‘const std::vector<fdeep::internal::tensor3>’   fdeep::tensor3_into_bytes(output_tensor,image2.data, image2.rows * image2.cols * image2.channels()^);
                                                                                                       
where i am doing mistake?

**Comment by Dobiasd** [comment]
[`model::predict`](https://github.com/Dobiasd/frugally-deep/blob/7d592b23e0124413f702db1f6a2e73d00320d721/include/fdeep/model.hpp#L19) does not return an `fdeep::tensor3` but an `std::vector<fdeep::tensor3>`. The reason for this is, that Keras models can also return more than one Tensor. But you can also simply see this in the functions signature or in your error message or in your IDE.

So in your case, this would help:
```cpp
const auto output_tensors = model.predict({tensor});
const auto output_tensor = output_tensors[0];
```

But since you seem to get confused by the types, I recommend you write them out explicitly instead of using `auto`, like so:
```cpp
const fdeep::tensor3s output_tensors = model.predict({tensor});
const fdeep::tensor3 output_tensor = output_tensors[0];
```

But in general your current problems do not seem to be rooted in the implementation of frugally-deep, but in fundamental problems in using C++. So instead of this issue here, stackoverflow (or a chat with some private programming tutor) would probably be a better place for questions like these.

**Comment by Huongninh** [comment]
> Ok, just created one example using [CImg](http://cimg.eu/) and one using [OpenCV](https://opencv.org/):
> 
>     * [cimg_example.cpp](https://gist.github.com/Dobiasd/21651861b73042762126e8eea52d9974)
> 
>     * [opencv_example.cpp](https://gist.github.com/Dobiasd/3140cfd9f539b6adb346e0b4a0ce157b)
> 
> 
> Also just [added them](https://github.com/Dobiasd/frugally-deep/commit/88493e8aeeaaef8ae66d2597b7b276cba8442224) to the README.
Hi Dobiasd, 
I already  used  **fdeep::tensor5_from_bytes** for create to use test for my trained model as your sample in OpenCV. 
   " const auto input = fdeep::tensor5_from_bytes((image.ptr(),
            static_cast<std::size_t>(image.rows),
            static_cast<std::size_t>(image.cols),
            static_cast<std::size_t>(image.channels()),
            0.0f, 1.0f); "
However It make an error "The model takes [(Just 1, Just 1, Just 100, Just 30, Just 1)] but provided was: [(1, 1, 30, 100, 1)]" 
Where's my mistake? Thank you!

**Comment by Dobiasd** [comment]
@Huongninh

> `The model takes [(Just 1, Just 1, Just 100, Just 30, Just 1)] but provided was: [(1, 1, 30, 100, 1)]`
> Where's my mistake?

`(1, 1, 100, 30, 1)`
is different from
`(1, 1, 30, 100, 1)`
